Vaadin Session/UI/Route Scopes Unavailable During Deserialization in UnserializableComponentWrapper

### Description

I've encountered an issue while handling non-serializable Vaadin components during session replication. To manage this, I wrapped the component using the new UnserializableComponentWrapper, mocking the serialization logic. Our goal is not to persist the component's state but to recreate it and perform some initialization tasks upon deserialization.

During the deserialization process, we attempt to retrieve the component from the Spring application context statically. However, this approach fails when the component depends on a bean with a vaadin-ui scope. At this stage, the Vaadin session, UI, and route scopes are not available, leading to an error.

#### Steps to Reproduce:

- Create a non-serializable Vaadin component with a dependency on a vaadin-ui scoped bean.
- Wrap the component with UnserializableComponentWrapper to bypass direct serialization.
- Attempt to deserialize the wrapper and retrieve the component from the Spring context.
- Observe the error due to the unavailability of Vaadin-specific scopes during deserialization.

#### Expected Behavior:

Vaadin session, UI, and route scopes should be available (or lazily initialized) during the deserialization process.

Components wrapped in UnserializableComponentWrapper with dependencies on Vaadin-scoped beans should be successfully recreated.

#### Actual Behavior:

Deserialization fails when accessing beans with vaadin-ui scope, as these scopes are not initialized yet.

<details><summary>Stack Trace Details</summary>
<p>

```
org.springframework.beans.factory.support.ScopeNotActiveException: Error creating bean with name 'vaadinUiScopeBean': Scope 'vaadin-ui' is not active for the current thread; consider defining a scoped proxy for this bean if you intend to refer to it from a singleton
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:385) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1443) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1353) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:904) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.ConstructorResolver.resolvePreparedArguments(ConstructorResolver.java:836) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:158) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1375) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1201) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:562) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:355) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:225) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveNamedBean(DefaultListableBeanFactory.java:1323) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveNamedBean(DefaultListableBeanFactory.java:1284) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveBean(DefaultListableBeanFactory.java:486) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:341) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:334) ~[spring-beans-6.1.15.jar:6.1.15]
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1261) ~[spring-context-6.1.15.jar:6.1.15]
	at org.vaadin.example.BeanUtil.getBean(BeanUtil.java:19) ~[classes/:na]
	at org.vaadin.example.NonSerializableComponentWrapper.deserialize(NonSerializableComponentWrapper.java:21) ~[classes/:na]
	at com.vaadin.kubernetes.starter.sessiontracker.UnserializableComponentWrapper.lambda$restoreComponent$0(UnserializableComponentWrapper.java:134) ~[kubernetes-kit-starter-2.4.1.jar:2.4.1]
	at java.base/java.util.Optional.ifPresent(Optional.java:178) ~[na:na]
	at com.vaadin.kubernetes.starter.sessiontracker.UnserializableComponentWrapper.restoreComponent(UnserializableComponentWrapper.java:129) ~[kubernetes-kit-starter-2.4.1.jar:2.4.1]
	at java.base/java.io.ObjectInputStream$ValidationList$1.run(ObjectInputStream.java:2801) ~[na:na]
	at java.base/java.io.ObjectInputStream$ValidationList$1.run(ObjectInputStream.java:2799) ~[na:na]
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:714) ~[na:na]
	at java.base/java.io.ObjectInputStream$ValidationList.doCallbacks(ObjectInputStream.java:2797) ~[na:na]
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:547) ~[na:na]
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:498) ~[na:na]
	at com.vaadin.kubernetes.starter.sessiontracker.serialization.TransientInjectableObjectInputStream.readWithTransients(TransientInjectableObjectInputStream.java:162) ~[kubernetes-kit-starter-2.4.1.jar:2.4.1]
	at com.vaadin.kubernetes.starter.sessiontracker.SessionSerializer.doDeserialize(SessionSerializer.java:600) ~[kubernetes-kit-starter-2.4.1.jar:2.4.1]
	at com.vaadin.kubernetes.starter.sessiontracker.SessionSerializer.deserialize(SessionSerializer.java:267) ~[kubernetes-kit-starter-2.4.1.jar:2.4.1]
	at com.vaadin.kubernetes.starter.sessiontracker.serialization.debug.SerializationDebugRequestHandler.tryDeserialize(SerializationDebugRequestHandler.java:328) ~[kubernetes-kit-starter-2.4.1.jar:2.4.1]
	at com.vaadin.kubernetes.starter.sessiontracker.serialization.debug.SerializationDebugRequestHandler.internalSerializeAndDeserialize(SerializationDebugRequestHandler.java:258) ~[kubernetes-kit-starter-2.4.1.jar:2.4.1]
	at com.vaadin.kubernetes.starter.sessiontracker.serialization.debug.SerializationDebugRequestHandler$Runner.accept(SerializationDebugRequestHandler.java:199) ~[kubernetes-kit-starter-2.4.1.jar:2.4.1]
	at com.vaadin.kubernetes.starter.sessiontracker.serialization.debug.SerializationDebugRequestHandler$Filter.doFilter(SerializationDebugRequestHandler.java:220) ~[kubernetes-kit-starter-2.4.1.jar:2.4.1]
	at jakarta.servlet.http.HttpFilter.doFilter(HttpFilter.java:53) ~[tomcat-embed-core-10.1.33.jar:6.0]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-6.1.15.jar:6.1.15]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.15.jar:6.1.15]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-6.1.15.jar:6.1.15]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.15.jar:6.1.15]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-6.1.15.jar:6.1.15]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.15.jar:6.1.15]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:397) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:905) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1741) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63) ~[tomcat-embed-core-10.1.33.jar:10.1.33]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]
Caused by: java.lang.IllegalStateException: No VaadinSession bound to current thread
	at com.vaadin.flow.spring.scopes.AbstractScope.getVaadinSession(AbstractScope.java:75) ~[vaadin-spring-24.6.6.jar:na]
	at com.vaadin.flow.spring.scopes.VaadinUIScope.getBeanStore(VaadinUIScope.java:110) ~[vaadin-spring-24.6.6.jar:na]
	at com.vaadin.flow.spring.scopes.AbstractScope.get(AbstractScope.java:44) ~[vaadin-spring-24.6.6.jar:na]
	at com.vaadin.flow.spring.scopes.VaadinUIScope.get(VaadinUIScope.java:41) ~[vaadin-spring-24.6.6.jar:na]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:373) ~[spring-beans-6.1.15.jar:6.1.15]
	... 68 common frames omitted
```

</p>
</details> 

Reproduction Repository:
https://github.com/myronenkom/kubernates-kit-issue

Environment:
Vaadin Version: 24.6.6
Kubernates Kit: 2.4.1

--- a/kubernetes-kit-starter/src/main/java/com/vaadin/kubernetes/starter/sessiontracker/UnserializableComponentWrapper.java
+++ b/kubernetes-kit-starter/src/main/java/com/vaadin/kubernetes/starter/sessiontracker/UnserializableComponentWrapper.java
@@ -144,6 +144,7 @@ public class UnserializableComponentWrapper<S extends Serializable, T extends Co
                     getElement().appendChild(component.getElement());
                 }
             } finally {
+                CurrentInstance.clearAll();
                 CurrentInstance.restoreInstances(instances);
                 cleaner.run();
             }

--- a/kubernetes-kit-starter/src/main/java/com/vaadin/kubernetes/starter/sessiontracker/UnserializableComponentWrapper.java
+++ b/kubernetes-kit-starter/src/main/java/com/vaadin/kubernetes/starter/sessiontracker/UnserializableComponentWrapper.java
@@ -12,6 +12,7 @@ package com.vaadin.kubernetes.starter.sessiontracker;
 import java.io.IOException;
 import java.io.Serial;
 import java.io.Serializable;
+import java.util.Map;
 import java.util.Objects;
 import java.util.function.Consumer;
 
@@ -21,7 +22,9 @@ import com.vaadin.flow.component.UI;
 import com.vaadin.flow.dom.Element;
 import com.vaadin.flow.dom.ElementUtil;
 import com.vaadin.flow.function.SerializableFunction;
+import com.vaadin.flow.internal.CurrentInstance;
 import com.vaadin.flow.internal.StateTree;
+import com.vaadin.flow.server.VaadinSession;
 
 /**
  * A wrapper component that allows an otherwise unserializable {@link Component}
@@ -126,7 +129,12 @@ public class UnserializableComponentWrapper<S extends Serializable, T extends Co
     }
 
     private void restoreComponent() {
-        getUI().map(UI::getSession).ifPresent(session -> {
+        getUI().ifPresent(ui -> {
+            VaadinSession session = ui.getSession();
+            Map<Class<?>, CurrentInstance> instances = CurrentInstance
+                    .getInstances();
+            CurrentInstance.set(UI.class, ui);
+            CurrentInstance.set(VaadinSession.class, session);
             Runnable cleaner = SessionUtil.injectLockIfNeeded(session);
             try {
                 getElement().removeAllChildren();
@@ -136,6 +144,7 @@ public class UnserializableComponentWrapper<S extends Serializable, T extends Co
                     getElement().appendChild(component.getElement());
                 }
             } finally {
+                CurrentInstance.restoreInstances(instances);
                 cleaner.run();
             }
         });

--- a/kubernetes-kit-starter/src/main/java/com/vaadin/kubernetes/starter/sessiontracker/serialization/TransientAwareHolder.java
+++ b/kubernetes-kit-starter/src/main/java/com/vaadin/kubernetes/starter/sessiontracker/serialization/TransientAwareHolder.java
@@ -89,6 +89,7 @@ final class TransientAwareHolder implements Serializable {
             runnable.run();
         } finally {
             if (instanceMap != null) {
+                CurrentInstance.clearAll();
                 CurrentInstance.restoreInstances(instanceMap);
                 cleaner.run();
             }

--- a/kubernetes-kit-starter/src/test/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializerTest.java
+++ b/kubernetes-kit-starter/src/test/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializerTest.java
@@ -24,12 +24,14 @@ import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.mockito.ArgumentCaptor;
+import org.mockito.MockedStatic;
 import org.springframework.mock.web.MockHttpSession;
 
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.component.Tag;
 import com.vaadin.flow.component.UI;
 import com.vaadin.flow.function.DeploymentConfiguration;
+import com.vaadin.flow.internal.CurrentInstance;
 import com.vaadin.flow.server.VaadinContext;
 import com.vaadin.flow.server.VaadinService;
 import com.vaadin.flow.server.VaadinServletService;
@@ -54,6 +56,7 @@ import static org.mockito.ArgumentMatchers.isNotNull;
 import static org.mockito.ArgumentMatchers.notNull;
 import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockStatic;
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.times;
@@ -232,16 +235,26 @@ class SessionSerializerTest {
         await().atMost(1000, MILLISECONDS).untilTrue(serializationCompleted);
         verify(serializationCallback).onSerializationSuccess();
 
-        try {
-            serializer.deserialize(infoList.get(0), httpSession);
-        } catch (Exception e) {
-            fail(e);
+        try (MockedStatic<CurrentInstance> mockedStatic = mockStatic(
+                CurrentInstance.class)) {
+            try {
+                serializer.deserialize(infoList.get(0), httpSession);
+            } catch (Exception e) {
+                fail(e);
+            }
+
+            mockedStatic.verify(() -> CurrentInstance.set(any(), any()),
+                    times(3));
+            mockedStatic.verify(() -> CurrentInstance.restoreInstances(any()),
+                    times(5));
+
+            verify(serializationCallback).onDeserializationSuccess();
+            Optional<Component> component = ui.getElement().getChildren()
+                    .toList().get(0).getChildren().toList().get(0)
+                    .getComponent();
+            assertThat(((Unserializable) component.get()).getName().fullName())
+                    .isEqualTo("Unserializable");
         }
-        verify(serializationCallback).onDeserializationSuccess();
-        Optional<Component> component = ui.getElement().getChildren().toList()
-                .get(0).getChildren().toList().get(0).getComponent();
-        assertThat(((Unserializable) component.get()).getName().fullName())
-                .isEqualTo("Unserializable");
     }
 
     @Test

--- a/kubernetes-kit-starter/src/test/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializerTest.java
+++ b/kubernetes-kit-starter/src/test/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializerTest.java
@@ -24,7 +24,6 @@ import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.mockito.ArgumentCaptor;
-import org.mockito.MockedStatic;
 import org.springframework.mock.web.MockHttpSession;
 
 import com.vaadin.flow.component.Component;
@@ -56,7 +55,6 @@ import static org.mockito.ArgumentMatchers.isNotNull;
 import static org.mockito.ArgumentMatchers.notNull;
 import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.mockStatic;
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.times;
@@ -233,28 +231,22 @@ class SessionSerializerTest {
 
         serializer.serialize(httpSession);
         await().atMost(1000, MILLISECONDS).untilTrue(serializationCompleted);
+        assertCurrentInstancesIsEmpty();
         verify(serializationCallback).onSerializationSuccess();
 
-        try (MockedStatic<CurrentInstance> mockedStatic = mockStatic(
-                CurrentInstance.class)) {
-            try {
-                serializer.deserialize(infoList.get(0), httpSession);
-            } catch (Exception e) {
-                fail(e);
-            }
-
-            mockedStatic.verify(() -> CurrentInstance.set(any(), any()),
-                    times(3));
-            mockedStatic.verify(() -> CurrentInstance.restoreInstances(any()),
-                    times(5));
-
-            verify(serializationCallback).onDeserializationSuccess();
-            Optional<Component> component = ui.getElement().getChildren()
-                    .toList().get(0).getChildren().toList().get(0)
-                    .getComponent();
-            assertThat(((Unserializable) component.get()).getName().fullName())
-                    .isEqualTo("Unserializable");
+        try {
+            serializer.deserialize(infoList.get(0), httpSession);
+        } catch (Exception e) {
+            fail(e);
         }
+
+        // TODO add this assert after the vaadin/flow#21277 change is available
+        // assertCurrentInstancesIsEmpty();
+        verify(serializationCallback).onDeserializationSuccess();
+        Optional<Component> component = ui.getElement().getChildren().toList()
+                .get(0).getChildren().toList().get(0).getComponent();
+        assertThat(((Unserializable) component.get()).getName().fullName())
+                .isEqualTo("Unserializable");
     }
 
     @Test
@@ -591,8 +583,30 @@ class SessionSerializerTest {
     private UnserializableComponentWrapper<State, Unserializable> createUnserializableComponentWrapper() {
         Unserializable unserializable = new Unserializable("Unserializable");
         return new UnserializableComponentWrapper<>(unserializable,
-                component -> new State(component.getName().fullName()),
-                state -> new Unserializable(state.text()));
+                SessionSerializerTest::unserializableComponentSerializer,
+                SessionSerializerTest::unserializableComponentDeserializer);
+    }
+
+    private static State unserializableComponentSerializer(
+            Unserializable unserializable) {
+        assertCurrentInstancesNotEmpty();
+        return new State(unserializable.getName().fullName());
+    }
+
+    private static Unserializable unserializableComponentDeserializer(
+            State state) {
+        assertCurrentInstancesNotEmpty();
+        return new Unserializable(state.text());
+    }
+
+    private static void assertCurrentInstancesNotEmpty() {
+        Assertions.assertNotNull(CurrentInstance.get(UI.class));
+        Assertions.assertNotNull(CurrentInstance.get(VaadinSession.class));
+    }
+
+    private static void assertCurrentInstancesIsEmpty() {
+        Assertions.assertNull(CurrentInstance.get(UI.class));
+        Assertions.assertNull(CurrentInstance.get(VaadinSession.class));
     }
 
     private static ConditionFactory await() {

--- a/kubernetes-kit-starter/src/test/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializerTest.java
+++ b/kubernetes-kit-starter/src/test/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializerTest.java
@@ -240,8 +240,7 @@ class SessionSerializerTest {
             fail(e);
         }
 
-        // TODO add this assert after the vaadin/flow#21277 change is available
-        // assertCurrentInstancesIsEmpty();
+        assertCurrentInstancesIsEmpty();
         verify(serializationCallback).onDeserializationSuccess();
         Optional<Component> component = ui.getElement().getChildren().toList()
                 .get(0).getChildren().toList().get(0).getComponent();

